[RNMobile] Add mobile Spacer component

lukewalczak opened the pull request:
## Description
Ref to gb-mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1436
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1428

**NOTE**: That PR requires `RangeControl` mobile component -> https://github.com/WordPress/gutenberg/pull/17282 to control spacer height using slider.

That PR adds a mobile `Spacer` component.

**UPDATES**:

* Implemented Accessibility  - double press on `Cell` is focusing the `Slider` once screen reader is enabled

* Corrected dark more [check the screenshot](https://user-images.githubusercontent.com/22746080/67369767-e3034100-f579-11e9-911d-8f8ed1b34e2a.png)

* Typing a value works properly [check the gif](https://user-images.githubusercontent.com/22746080/67405004-70fc1d80-f5b4-11e9-8878-b98d1f4d8f21.gif)

* Styling changes according to @iamthomasbishop comment:
  * outline border-width is set to 2px
  * outline border color is set to blue-30
  * dashed border color is dimmed  (30%)
  * added padding-bottom to cell (16px)
  * added `done` button

iOS: Screenshot [one](https://user-images.githubusercontent.com/22746080/67470800-a6554980-f64e-11e9-960f-c5964dcb877c.PNG) and [two](https://user-images.githubusercontent.com/22746080/67470808-a9503a00-f64e-11e9-9544-3569e59d4c64.PNG)
Android: Screenshot [one](https://user-images.githubusercontent.com/22746080/67513376-fce87500-f69a-11e9-8b10-92f8f860670c.png) and [two](https://user-images.githubusercontent.com/22746080/67513379-feb23880-f69a-11e9-87e0-f73726868f6b.png).

* Fix focusing slider when VO is ON on iPhoneX

* Add an information for VO about current value when changing value



## How has this been tested?

* Choose `Spacer` component.
* Press settings button
* Play with a spacer height using slider:
  * default value: 100
  * minValue: 20
  * maxValue: 500 **NOTE:** if user sets height 800 on the website and then edits it on the phone, the value from web can be in the middle of the range, so user can decrease and increase the value as well. 
* Save post on mobile
* Open the same post on the web
* Play with the height, e.g. set it to 1000 and save a post
* Open the same post again on mobile and check the `Spacer` height

## Screenshots <!-- if applicable -->

android | ios
--- | ---
![spacer_android](https://user-images.githubusercontent.com/22746080/66636940-6b94ef80-ec12-11e9-89af-34cd1b505925.gif) | ![spacer_ios](https://user-images.githubusercontent.com/22746080/66636950-6f287680-ec12-11e9-880d-7c2414999047.gif)

* setting a spacer height on web

![spacer_4](https://user-images.githubusercontent.com/22746080/67196269-c5f43400-f3fa-11e9-8db7-cff10caaa12f.gif)

* setting a spacer height on mobile

![spacer_5](https://user-images.githubusercontent.com/22746080/67196545-53378880-f3fb-11e9-92fa-357150149f1c.gif)


**NOTE:** On the gifs there is a wrong label `Slider` instead of `Height in pixels` which is correct  in the code.


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

lukewalczak committed:
Revert package-lock.json

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

gziolo commented on packages/block-library/src/spacer/edit.native.js:
```diff
@@ -0,0 +1,62 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
```
You don't need to explicitly import `React`. It's handled with the Babel preset.

gziolo commented on packages/block-library/src/spacer/edit.native.js:
```diff
@@ -0,0 +1,62 @@
+		} return setSpacerHeight( value );
+	};
+
+	const getInspectorControls = () => (
```
This can be either inlined into the `View` or assigned directly to the const:
```suggestion
	const inspectorControls = (
```

lukewalczak replied:
I would like to keep it outside a `View`,  because on the next iteration once we will add a draggable functionality in can mess a code a bit. Assigning it directly to the const makes sense, thanks!

lukewalczak replied:
I've decided to move it into `View` finally.

lukewalczak committed:
Setting attributes for spacer height

lukewalczak committed:
Correct the condition for setting maximum value

lukewalczak commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -20,7 +20,7 @@ export default function BottomSheetRangeCell( props ) {
 		minimumTrackTintColor = '#00669b',
 		maximumTrackTintColor = Platform.OS === 'ios' ? '#e9eff3' : '#909090',
-		thumbTintColor = Platform.OS === 'ios' ? '#fff' : '#00669b',
+		thumbTintColor = Platform.OS === 'android' && '#00669b',
```
I've noticed that if `thumbTintColor` on iOS is set to white color it losing the shadow on iPhoneX, however it works as expected on other iPhones. I've decided to use the custom color 

with white color | with default color
--- | ---
![IMG_0008](https://user-images.githubusercontent.com/22746080/67198083-7fa0d400-f3fe-11e9-8a13-04dace43e1ce.PNG) | ![IMG_0007](https://user-images.githubusercontent.com/22746080/67198087-80d20100-f3fe-11e9-81c7-1f3283de7ed1.PNG)

Currently we are using the `Slider` from react-native, which will be deprecated soon and we'll have to use the `Slider` from community library. Need to check if the bug doesn't exist in that library.

lukewalczak replied:
I've tested this using slider library from a community and it seems like a bug on ios version ~12.

Tested on:
* device - iPhone X - 12.3.1 - doesn't work properly
* device - iPhone X - 13.1.3 - works as expected
* device - iPhone 7 - 12.4 -  doesn't work properly
* device -  iPhone 7 - 13.1.3  - works as expected

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

etoledom reviewed (commented):
Hey @lukewalczak - Thank you for this PR!

I added a few small code comments. 

About DarkMode, if we want the box to be black, I think the best would be to just not set a color to the background. cc @iamthomasbishop

<img src="https://user-images.githubusercontent.com/9772967/67278148-b4795d80-f4c8-11e9-8881-8f6eec9dd294.PNG" width="300">

Some other few details:

- The block size is not updating while the slider is moved. I don't think this is a problem but just remarking it since it's different from the given gifs on the description.
![size](https://user-images.githubusercontent.com/9772967/67278227-dd99ee00-f4c8-11e9-8ecc-50ec91d19cae.gif)

- Setting the size via the input field won't modify the setting value:
![setting-size](https://user-images.githubusercontent.com/9772967/67278483-58fb9f80-f4c9-11e9-8f2a-6f938d158695.gif)

- On Android, if the label is just a bit longer, there's no space left for the slider. Probably we will need to have the slider control on a second line:
![Screenshot_20191022-123320](https://user-images.githubusercontent.com/9772967/67278702-c60f3500-f4c9-11e9-93c4-cc1834898b0c.png)

- I think we should also improve **Accessibility**. Currently the control is inaccessible to VoiceOver on iOS. We could make the controls (slider and input field) accessible after the user double-taps on the cell, and automatically select the slider with (maybe) its native accessibility description.

Probably an accessibility label on the cell itself will also help, It would be great if it let the user know the current value without much effort.

- About the design, whatever @iamthomasbishop says. I'm just wondering if we could do something about the input field.
Maybe we could center the text and change the border color? Is it possible to have just the bottom border like the original design?

lukewalczak commented on packages/block-library/src/spacer/editor.native.scss:
```diff
@@ -0,0 +1,10 @@
+.staticSpacer {
+	height: 20px;
+	background-color: $white;
```
Setting to `transparent` will be ok. Thanks!

lukewalczak commented on packages/block-library/src/spacer/edit.native.js:
```diff
@@ -0,0 +1,71 @@
+	return (
+		<View style={ [ styles.staticSpacer, isSelected && styles.selectedSpacer, { height } ] }>
+			<InspectorControls>
+				<PanelBody title={ __( 'SpacerSettings' ) } >
```
👍

lukewalczak commented:
Thanks @etoledom for you valuable feedback 👍 . I would like to refer to your comments:

> About DarkMode, if we want the box to be black, I think the best would be to just not set a color to the background.

I will set the background color as a `transparent` as you've suggested. It will look like:
<img width="300" alt="Screenshot 2019-10-23 at 09 44 54" src="https://user-images.githubusercontent.com/22746080/67369767-e3034100-f579-11e9-911d-8f8ed1b34e2a.png">

Please @iamthomasbishop let us know if this is correct.

> The block size is not updating while the slider is moved.

That's true, value is not updating live when slider is moved, because value is saved on `onSlidingComplete` inside a `Slider` component. As far as I remember @jbinda did this because of input validation.

> Setting the size via the input field won't modify the setting value

I was surprised that it doesn't work, so I tested it before I had started debugging and I discovered that `onBlur` in `TextInput` next to the `Slider` is not called once closing bottom sheet. Need to dive into it.

> On Android, if the label is just a bit longer, there's no space left for the slider. Probably we will need to have the slider control on a second line

> About the design. I'm just wondering if we could do something about the input field.

Generally, the goal is to have a slider control on a second line. Also new styling is already planned. There is an [issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1448) for that.

> I think we should also improve Accessibility.

Definitely, will test it and do what we need to achieve in that matter.

lukewalczak committed:
Small code refactor

lukewalczak committed:
Improve Accessibility in range-cell

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

lukewalczak committed:
More accessibility improvements

lukewalczak commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -20,14 +25,38 @@ export default function BottomSheetRangeCell( props ) {
+			const reactTag = findNodeHandle( this.sliderRef );
+			AccessibilityInfo.setAccessibilityFocus( reactTag );
+		}
+	};
```
* What I wanted to achieve here is to make the parent component (`Cell`) not accessible, since I want to focus `Slider` and have also `TextInput` accessible as well. 

* Wdyt about having there only one component when screen reader is enabled?

* Both platforms are reading the value in percents, e.g. `100` px is is read as `17 percents`.

lukewalczak replied:
@etoledom Could you please look at it? I would like to know if something is wrong there. Thanks!

pinarol commented:
> I think we should also improve Accessibility. Currently the control is inaccessible to VoiceOver on iOS. We could make the controls (slider and input field) accessible after the user double-taps on the cell, and automatically select the slider with (maybe) its native accessibility description.

Super catch! We can handle this on this [issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1448) as well. We won't be shipping to prod without solving that issue.

lukewalczak commented:
I've already implemented a11y changes within that PR.

lukewalczak commented on packages/block-library/src/index.native.js:
```diff
@@ -144,6 +144,7 @@ export const registerCoreBlocks = () => {
 		mediaText,
 		// eslint-disable-next-line no-undef
 		!! __DEV__ ? group : null,
+		spacer,
```
Should it be able only for dev currently as group?

pinarol replied:
Yes 👍 We'll keep it that way until we finish Slider improvements

pinarol commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -20,14 +25,38 @@ export default function BottomSheetRangeCell( props ) {
+
+	const accessibilityLabel =
+	sprintf(
+		/* translators: accessibility text. Inform about current spacer height value. %2$s: Spacer height current value. */
```
RangeCell is not limited to Spacer block so we are not supposed to put block specific description here.

lukewalczak replied:
Make sense, thanks!

pinarol commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -20,14 +25,38 @@ export default function BottomSheetRangeCell( props ) {
+	const accessibilityLabel =
+	sprintf(
+		/* translators: accessibility text. Inform about current spacer height value. %2$s: Spacer height current value. */
+		_x( '%1$s. Current value is %2$s px', 'range cell' ),
```
Similarly not everything we adjust with a Slider will be in pixels, so I suggest removing px. 

and 'range cell' might not be an ideal description for translators. I am thinking sth like 'Slider for picking a number inside a range' ?

lukewalczak replied:
To be honest I was not sure what should be there, so I will accept your suggestion 👍

lukewalczak committed:
Small code refactor

iamthomasbishop commented:
@lukewalczak This is looking pretty solid, well done! I have a few bits of feedback, and I'll update my design library + zeplin based on this discussion.

* In dark mode, we can probably dim the dotted border a bit, something like white at 30% alpha.

* Can we use the slightly lighter shade of blue on the selection outline? The same color we've been using for links, which is `blue 30` from Color Studio, aka `#5198d9`.

* Is the `border-width` on the selection 1px? Can we bump slightly wider to match other selection styling (see Video block for reference) which I think is 2px.

* Can we apply either 16px padding at the bottom of the inspector sheet so there is a little breathing room? (Note: If 16px is still too shallow, we can go with 24px)

One other thing that we might want to consider (maybe separate from this ticket) is that we might want to provide a `Done` button of sorts to apply the value of TextInputs. Example:

<img width="918" alt="Screen Shot 2019-10-23 at 11 29 53 AM" src="https://user-images.githubusercontent.com/1200432/67414372-93784180-f588-11e9-84f4-a9a897787a30.png">

(left uses our existing Quick Toolbar style, right uses a more "system" style toolbar)

lukewalczak commented:
@iamthomasbishop thank you! Could you please let me know:

> Is the border-width on the selection 1px? Can we bump slightly wider to match other selection styling (see Video block for reference) which I think is 2px.

Do you mean text input border-width? If yes, maybe it's better to track it within issue related to [range cell styling](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1448).

> One other thing that we might want to consider (maybe separate from this ticket) is that we might want to provide a Done button of sorts to apply the value of TextInputs

I will check what we can to in that matter!

iamthomasbishop commented:
> Correct if I'm wrong but probably there should be dashed border

You're right, I was using the `border-style` name 🤦‍♂ 

> Do you mean selection outline of text input?

Sorry for not being clearer. I was referring to the selection outline the block body gets when selected (example: https://cloudup.com/c6_z8XNNjNL)

koke commented:
This is looking pretty good ✨ . I haven't done a thorough review, but I spotted a few things:

- When you use the text input, the slider updates as you type but the spacer doesn't.
- At least on iOS, I don't see how to dismiss the keyboard without interacting with the slider or dismissing the block settings.
- The slider's thumb seems a bit jiggly as you drag it sometimes:

![Screen Recording 2019-10-23 at 23 06 30](https://user-images.githubusercontent.com/8739/67434257-0fef3c80-f5ea-11e9-96a1-ee5e6521658b.gif)

As for the code:

- I would keep only the slider in `Slider` and move the `TextInput` to `RangeCell`
- Similarly, I think the `changeSpacerHeight` logic in `SpacerEdit` really belongs in `RangeCell`. We are already passing a minimum and maximum value to it, and it should be responsible to coordinate its two entry controls.

lukewalczak committed:
Styling spacer refactor

lukewalczak committed:
Move logic to RangeCell

lukewalczak commented:
@koke I've extracted logic in `RangeCell` as you requested. I'm still wondering what will be an advantage of moving `TextInput` into `RangeCell` instead having it inside a `Slider`. Could you please elaborate on that?

Also after the refactor I was not able to reproduce that 'jiggly' behaviour.

lukewalczak committed:
Keep Slider along with TextInput within RangeCell

lukewalczak committed:
Small cleanup

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

iamthomasbishop commented:
@lukewalczak Looks nice in the updated screenshots, thanks for wrangling those changes!

> dashed border color is dimmed (30%)

Just want to confirm, this is only on dark mode, correct?

> added padding-bottom to cell (16px)

This looks better, but can we remove the trailing separator on the last table row?

> added done button

This is great! Does this also get an Android equivalent? 

Two more things, neither of which are blockers:

* If it's not a hassle, can we make the `Done` text label `blue 30` as well?
* Regarding the Slider component in general, can we force the Slider/TextInput "group" to a second line in the same table row? This is a bit narrow.

koke commented:
> EDIT: Maybe we don't need a `Slider` component itself (as a separate component)? - we can have a `Slider` along with `TextInput` inside `RangeCell`. Is that what you meant? - I've refactored it actually in that way, let me know wdyt.

That looks like what I had in mind, but for some reason I thought the Slider part had some more custom behavior than adding the input.

lukewalczak commented:
@koke Do you think we can leave the structure as it is right now?

lukewalczak commented:
@iamthomasbishop I've updated PR description and added Android screenshots 🎉 

> Regarding the Slider component in general, can we force the Slider/TextInput "group" to a second line in the same table row? This is a bit narrow

Generally, that issue will be fix within [styling](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1448) issue, but honestly I'm all for doing it within in this PR. Just let me know if you are ok with that ✌️.

iamthomasbishop commented:
> Generally, that issue will be fix within styling issue, but honestly I'm all for doing it within in this PR. Just let me know if you are ok with that ✌️.

@lukewalczak Ah ok, either way is fine by me although the sooner the better 😄

lukewalczak committed:
Fix missing binds

etoledom reviewed (commented):
Thank you for the improvements @lukewalczak !

I gave it a try with VoiceOver on device (iPhone X), and double-tapping on the cell didn't focus the slider. Did a bit of debugging and I noticed that `onCellPress` is not being called.
It is called when I tap on it normally, but it won't with VoiceOver.

About: https://github.com/WordPress/gutenberg/pull/17896#discussion_r338066598

> What I wanted to achieve here is to make the parent component (Cell) not accessible, since I want to focus Slider and have also TextInput accessible as well.

That sounds good! It's the approach we follow on other blocks too 👍 

> Wdyt about having there only one component when screen reader is enabled?

Could you please expand a bit? Do you mean having only slider?
If that's the case, wouldn't auto-focusing the slider be enough and simpler?

> Both platforms are reading the slider value in percents, e.g. 100 px is is read as 17 percents.

True that, and it's probably not a good thing in this case. Since 10% can mean different thing on different situations, and in this case, there's no visual help for the user. I think it's best if we can let them know the precise height amount. 

A good example is the Music app. It has the the Volume scroller that says its value on percent, and there's the track position scroller. Since songs has different lengths, it won't make much sense to say percents there, so they say the current time together with the total time of the song while dragging it.

If this is blocking other work, I'm happy to ✅ as it is and improve these details separately.

Everything else looks great to me! 🎉 
Thank you again!

lukewalczak commented:
@etoledom Thanks for feedback! I was testing accessibility on iPhone 7, so will dive into the issue on iPhoneX.

> A good example is the Music app. It has the the Volume scroller that says its value on percent, and there's the track position scroller. Since songs has different lengths, it won't make much sense to say percents there, so they say the current time together with the total time of the song while dragging it.

I'm not sure what can I do in that matter. Probably I can inform user about current value when user releases the finger, however voice over will also pass the info in percents.

lukewalczak committed:
Fix focusing slider on iphoneX when VO is on

lukewalczak committed:
Adjust a11y voice over

lukewalczak commented:
@etoledom Could you please re-test it? 🙏

lukewalczak committed:
Refactor pointerEvents when screen reader is on

lukewalczak committed:
Announce current value when finished

lukewalczak committed:
Improve a11y

lukewalczak committed:
Fix a11y on iPhoneX

etoledom commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -1,46 +1,180 @@
+			/* translators: accessibility text. Inform about current value. %2$s: Current value. */
+			_x( '%1$s. Current value is %2$s', 'Slider for picking a number inside a range' ),
+			cellProps.label, value
+		);
```
It would probably be good to add what `%1$s` is for.
This text is the guide and (usually) only context for translators to do a good translation work. The better context, the better translations we get :)

Something like: `%1$s: Control label` or similar should help 👍

lukewalczak replied:
Added! Thanks!

etoledom reviewed (approved):
Accessibility on iOS is working better now. Thanks @lukewalczak for the update!

I added a small code comment, but overall looks good to me 🎉 
Great job 🙏

lukewalczak committed:
Update info for translators
